### PR TITLE
[Feature] Add quota on billing group for resources tracking

### PR DESCRIPTION
**Pull request by EliasBoulharts**

### CODE REVIEW

**Comment by EliasBoulharts** [comment]
TODO:

File: profiles/models/quota_binding.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    @property
+    def available(self):
+        return self.limit - self.consumed
+
+    @property
+    def percentage(self):
+        if self.limit == 0:
+            return None
+        percent_consumed = 0
+        try:
```
Comment (Sispheor) [inline profiles/models/quota_binding.py:38]: as the limit is tested before. no needs to try catch here

File: profiles/tables/billing_group_table.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+
+class BillingGroupTable(tables.Table):
+    name = LinkColumn("profiles:billing_group_details", args=[A("id")])
+    actions = TemplateColumn(template_name='custom_columns/group_actions.html', orderable=False)
+
+    class Meta:
+        model = BillingGroup
+        attrs = {"id": "billing_group_table", "class": "table squest-pagination-tables "}
+        fields = ("name", "actions")
```
Comment (Sispheor) [inline profiles/tables/billing_group_table.py:14]: view to be removed

File: profiles/views/quota_attribute_definition_views.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    breadcrumbs = [
+        {'text': 'Quota attributes', 'url': reverse('profiles:quota_attribute_definition_list')},
+        {'text': quota_attribute_definition.name, 'url': ""}
+    ]
+    context = {
+        'breadcrumbs': breadcrumbs,
+        'confirm_text': mark_safe(f"Confirm deletion of <strong>{quota_attribute_definition.name}</strong>?"),
+        'action_url': reverse('profiles:quota_attribute_definition_delete', kwargs=args),
+        'button_text': 'Delete',
+        'details': {'warning_sentence': 'Warning: some users are still present in this quota_attribute_definition:',
```
Comment (Sispheor) [inline profiles/views/quota_attribute_definition_views.py:84]: Not "users" but attribute definition

Comment (Sispheor) [reply profiles/views/quota_attribute_definition_views.py:84]: This quota attribute was linked to the following billing groups (list of billing group)

File: templates/profiles/billing-group-details.html [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                    <li class="nav-item">
+                        <a class="nav-link" href="#users" data-toggle="tab">Users</a>
+                    </li>
+                </ul>
+            </div>
+            <div class="card-body">
+                <div class="tab-content">
+                    <div class="active tab-pane" id="quota">
+                        <a class="btn btn-success float-sm-right" href="{% url "profiles:quota_binding_edit_all" object_id %}">
+                            <i class="fab fa-cloudscale"></i> Manage attribute and limits
```
Comment (Sispheor) [inline templates/profiles/billing-group-details.html:37]: Update the icon with: `<i class="far fa-chart-bar"></i>`

File: templates/sidebar.html [hunk 5]
```diff
@@ -90,6 +90,31 @@
...
+                                    <p>Groups</p>
+                                </a>
+                            </li>
+                            <li class="nav-item">
+                                <a href="{% url 'profiles:quota_attribute_definition_list' %}"
+                                   class="nav-link {% if '/quota-attribute/' in request.path %}active{% endif %}">
+                                    <i class="nav-icon far fa-chart-bar"></i>
+                                    <p>Quota</p>
+                                </a>
+                            </li>
```
Comment (Sispheor) [inline templates/sidebar.html:115]: Place the quota block before groups as we need to configure this one first.
+ change icon for `<i class="fas fa-chart-pie"></i>`

File: profiles/views/quota_attribute_definition_views.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+from django.urls import reverse
+from django.utils.safestring import mark_safe
+
+from profiles.filters.quota_attribute_definition_filter import QuotaAttributeDefinitionFilter
+from profiles.forms.quota_attribute_definition_form import QuotaAttributeDefinitionForm
+from profiles.models import QuotaAttributeDefinition
+from profiles.tables.quota_attribute_definition_table import QuotaAttributeDefinitionTable
+
+
+class QuotaAttributeDefinitionListView(LoginRequiredMixin, SingleTableMixin, FilterView):
```
Comment (Sispheor) [inline profiles/views/quota_attribute_definition_views.py:15]: admin protected page

File: tests/test_profile/test_quota/test_model.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+from service_catalog.models import Instance
+from tests.test_profile.test_quota.base_test_quota import BaseTestQuota
+
+
+class TestQuotaModel(BaseTestQuota):
+
+    def setUp(self):
+        super(TestQuotaModel, self).setUp()
+
+    def test_consumed_updated_after_billing_group_changed_in_instance(self):
```
Comment (Sispheor) [inline tests/test_profile/test_quota/test_model.py:11]: check resource consumption data. We remove 30 vcpu fron one side and we retrieve them on the other side

File: tests/test_profile/test_quota/test_model.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        self.assertNotEqual(old_consumed_cpu_old_billing_group, old_billing_group.quota_bindings.get(
+            quota_attribute_definition=self.test_quota_attribute_cpu).consumed)
+        self.assertNotEqual(old_consumed_memory_old_billing_group, old_billing_group.quota_bindings.get(
+            quota_attribute_definition=self.test_quota_attribute_memory).consumed)
+        self.assertNotEqual(old_consumed_cpu_new_billing_group, new_billing_group.quota_bindings.get(
+            quota_attribute_definition=self.test_quota_attribute_cpu).consumed)
+        self.assertNotEqual(old_consumed_memory_new_billing_group, new_billing_group.quota_bindings.get(
+            quota_attribute_definition=self.test_quota_attribute_memory).consumed)
+
+    def test_consumed_updated_after_attribute_definitions_changed_in_quota_attribute(self):
```
Comment (Sispheor) [inline tests/test_profile/test_quota/test_model.py:39]: check delta between values

File: profiles/models/quota_binding.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    quota_attribute_definition = ForeignKey(
+        to=QuotaAttributeDefinition,
+        blank=False,
+        null=False,
+        on_delete=CASCADE,
+        related_name='quota_bindings'
+    )
+
+    @property
+    def available(self):
```
Comment (Sispheor) [inline profiles/models/quota_binding.py:30]: add test

File: profiles/models/quota_binding.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        return self.limit - self.consumed
+
+    @property
+    def percentage(self):
```
Comment (Sispheor) [inline profiles/models/quota_binding.py:34]: add test for all cases

File: profiles/models/quota_binding.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        if self.limit == 0:
+            return None
+        percent_consumed = (self.consumed * 100) / self.limit
+        return round(percent_consumed)
+
+    def update_consumed(self):
```
Comment (Sispheor) [inline profiles/models/quota_binding.py:40]: assert called on update

**Comment by Sispheor** [comment]
Missing tests:

- [ ] assert called on each signal (from all attached object)
- [ ] test form when overridden save method and/or init method
- [ ] test views (POST) create and edit page

**Review by Sispheor — CHANGES_REQUESTED** [review]
Some test to be added. Rest is ok.

File: tests/test_profile/test_group/test_group_urls.py [hunk 12]
```diff
@@ -7,6 +7,7 @@ class TestGroupUrls(TestGroupBase):
 
     def setUp(self):
         super(TestGroupUrls, self).setUp()
+        self.client.force_login(self.superuser)
```
Comment (Sispheor) [inline tests/test_profile/test_group/test_group_urls.py:10]: to be removed

File: tests/test_profile/test_quota/test_model.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+class TestQuotaModel(BaseTestQuota):
+
+    def setUp(self):
+        super(TestQuotaModel, self).setUp()
+
+    def test_get_available(self):
+        quota_binding = self.test_billing_group.quota_bindings.first()
+        quota_binding.limit = 100
+        self.assertEqual(quota_binding.available, 100 - quota_binding.consumed)
```
Comment (Sispheor) [inline tests/test_profile/test_quota/test_model.py:14]: add a test on the "consumed" property

File: tests/test_profile/test_quota/test_urls.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        self._check_post_page(reverse('profiles:quota_binding_edit_all', kwargs=self.args_billing),
+                              data={'quota_attribute_definition': [self.test_quota_attribute_cpu.id]})
+        self._check_post_page(url, data={self.test_quota_attribute_cpu.name: 50})
+        binding = self.test_billing_group.quota_bindings.get(quota_attribute_definition=self.test_quota_attribute_cpu)
+        self.assertEqual(binding.limit, 50)
+        self.client.logout()
+        self._check_get_page(url, 302)
+
+    def test_quota_attribute_definition_list(self):
+        url = reverse('profiles:quota_attribute_definition_list')
```
Comment (Sispheor) [inline tests/test_profile/test_quota/test_urls.py:67]: add a check on returned data

**Review by Sispheor — CHANGES_REQUESTED** [review]
More tests to be added.

File: tests/test_profile/test_quota/test_model.py [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+
+    def test_get_percentage(self):
+        quota_binding = self.test_billing_group.quota_bindings.first()
+        quota_binding.limit = 100
+        self.assertEqual(quota_binding.percentage, quota_binding.consumed)
+
+    def test_update_consumed(self):
+        consumed = 0
+        for instance in self.test_billing_group.instances.all():
```
Comment (Sispheor) [inline tests/test_profile/test_quota/test_model.py:23]: add local object with hard coded values and then check

**Review by Sispheor — CHANGES_REQUESTED** [review]
Ultimate last test to be added

**Pull request merged**